odl/trafos/util/ft_utils.py

```diff
@@ -0,0 +1,654 @@
+        non-floating type, it is converted to 'float64'.
+    shift : bool or or `sequence` of bools, optional
+        If ``True``, the grid is shifted by half a stride in the negative
+        direction. With a sequence, this option is applied separately on
+        each axis.
+    axes : int or `sequence` of ints, optional
+        Dimensions in which to calculate the reciprocal. The sequence
+        must have the same length as ``shift`` if the latter is given
+        as a sequence.
+        Default: all axes.
+    sign : {'-', '+'}, optional
+        Sign of the complex exponent.
+    out : `numpy.ndarray`, optional
+        Array in which the result is stored. If ``out is arr``,
+        an in-place modification is performed. For real data type,
+        this is only possible for ``shift=True`` since the factors are
+        complex otherwise.
+
+    Returns
+    -------
+    out : `numpy.ndarray`
+        Result of the pre-processing. If ``out`` was given, the returned
+        object is a reference to it.
+
+    Notes
+    -----
+    If ``out`` is not specified, the data type of the returned array
+    is the same as that of ``arr`` except when ``arr`` has real data
+    type and ``shift`` is not ``True``. In this case, the return type
+    is the complex counterpart of ``arr.dtype``.
+    """
+    arr = np.asarray(arr)
+    if not is_scalar_dtype(arr.dtype):
+        raise ValueError('array has non-scalar data type {}'
+                         ''.format(dtype_repr(arr.dtype)))
+    elif is_real_dtype(arr.dtype) and not is_real_floating_dtype(arr.dtype):
+        arr = arr.astype('float64')
+
+    if axes is None:
+        axes = list(range(arr.ndim))
+    else:
+        try:
+            axes = [int(axes)]
+        except TypeError:
+            axes = list(axes)
+
+    shape = arr.shape
+    shift_list = normalized_scalar_param_list(shift, length=len(axes),
+                                              param_conv=bool)
+
+    # Make a copy of arr with correct data type if necessary, or copy values.
+    if out is None:
+        if is_real_dtype(arr.dtype) and not all(shift_list):
+            out = np.array(arr, dtype=TYPE_MAP_R2C[arr.dtype], copy=True)
+        else:
+            out = arr.copy()
+    elif out is arr:
+        pass
+    else:
+        out[:] = arr
+
+    if is_real_dtype(out.dtype) and not shift:
+        raise ValueError('cannot pre-process real input in-place without '
+                         'shift')
+
+    if sign == '-':
+        imag = -1j
+    elif sign == '+':
+        imag = 1j
+    else:
+        raise ValueError("`sign` '{}' not understood".format(sign))
+
+    def _onedim_arr(length, shift):
+        if shift:
+            # (-1)^indices
+            indices = np.arange(length)
+            arr = -2 * np.mod(indices, 2) + 1.0
+        else:
+            indices = np.arange(length)
+            arr = np.exp(-imag * np.pi * indices * (1 - 1.0 / length))
+        return arr.astype(out.dtype, copy=False)
```

Comment:
Create a arange of the correct dtype from the start instead, then you can avoid this and optimize above:

```
arr = np.arange(length)
np.exp((-imag * np.pi  * (1 - 1.0 / length)) * arr, out=arr)
```